tests/jobs/test_local_task_job.py

```diff
@@ -81,6 +81,110 @@ def test_localtaskjob_essential_attr(self):
+                                     executor=SequentialExecutor())
+            valid_job.run()
+            valid_ti = dr.get_task_instance(task_id=valid_task.task_id,
+                                            session=session)
+            # task that returns valid return code should succeed.
+            self.assertEqual(valid_ti.state, State.SUCCESS)
+            self.assertIsNotNone(valid_ti.pid)
+
+    def test_invalid_return_code_with_python_operator(self):
+        with create_session() as session:
+            dagbag = models.DagBag(
+                dag_folder=TEST_DAG_FOLDER,
+                include_examples=False,
+            )
+            dag_id = 'test_invalid_return_code'
+            valid_task_id = 'test_valid_return_code_python_operator'
+            invalid_task_id = 'test_invalid_return_code_python_operator'
+            dag = dagbag.get_dag(dag_id)
+            invalid_task = dag.get_task(invalid_task_id)
+            dr = dag.create_dagrun(run_id="test",
+                                   state=State.SUCCESS,
+                                   execution_date=DEFAULT_DATE,
+                                   start_date=DEFAULT_DATE,
+                                   session=session)
+            invalid_ti = TI(task=invalid_task, execution_date=DEFAULT_DATE)
+            invalid_ti.refresh_from_db()
+            invalid_ti.state = State.RUNNING
+            invalid_ti.hostname = get_hostname()
+            session.commit()
+
+            invalid_job = LocalTaskJob(task_instance=invalid_ti,
+                                       ignore_ti_state=True,
+                                       executor=SequentialExecutor())
+            with self.assertRaisesRegex(AirflowException,
+                                        "LocalTaskJob process exited with non zero status"):
+                invalid_job.run()
+
+            invalid_ti = dr.get_task_instance(task_id=invalid_task.task_id,
+                                              session=session)
+            # task that returns invalid return code should fail.
+            self.assertEqual(invalid_ti.state, State.FAILED)
+            self.assertIsNotNone(invalid_ti.pid)
+
+            valid_task = dag.get_task(valid_task_id)
+            valid_ti = TI(task=valid_task, execution_date=DEFAULT_DATE)
+            valid_ti.refresh_from_db()
+            valid_ti.state = State.RUNNING
+            valid_ti.hostname = get_hostname()
+            session.commit()
+
+            valid_job = LocalTaskJob(task_instance=valid_ti,
+                                     ignore_ti_state=True,
+                                     executor=SequentialExecutor())
+            valid_job.run()
+            valid_ti = dr.get_task_instance(task_id=valid_task.task_id,
+                                            session=session)
+            # task that returns valid return code should succeed.
+            self.assertEqual(valid_ti.state, State.RUNNING)
```

Comment:
If the job has run why is this TI still in running state? It should be succeeded, no?

Reply:
That's the problem of python operator.....if you call sys.exit(0) from the python operator. It will not do any post processing like setting the state to succeed which is done in python operator. 

So for this case I have following suggestions
1. disallow calling sys.exit in python operator, this can be done by a code scan before running.
2. running the python operator in another process and having the task state management in its own process.

I think it will be another PR if we really want to fix this.

Reply:
Ah. 
Ignoring `os._exit()` for now, `sys.exit` could be handled in the python operator by catching SystemExit exception. (That would be a separate PR though)

Reply:
Oh, yep, I mean os._exit

Reply:
Ah okay.

But with your new change shouldn't the on_failure have set the TI to failed?

Reply:
i.e. if we're checking the exit code shouldn't we have noticed that the task state hasn't updated and mark it as failed as a result? (with a log message so the reason for the failure shows up in the task logs)

Reply:
> Ah okay.
> 
> But with your new change shouldn't the on_failure have set the TI to failed?

In my change on_failure will only be triggered when the return code is non zero.  Here we called os._exit(0)

Reply:
> i.e. if we're checking the exit code shouldn't we have noticed that the task state hasn't updated and mark it as failed as a result? (with a log message so the reason for the failure shows up in the task logs)

It's not a failure, the exit code is zero. Maybe we can add another check that 
if we get a return code and the state is not in State.finished(), we set the task state accordingly. I can make that change as well but it seems its out of the scope of this PR.

Reply:
You aren't checking the state of the job anywhere.

Reply:
do you mean I need to check the state of the job in the unittest? Just added one